ns/Cns_stat.man
.\" Copyright (C) 1999-2003 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH CNS_STAT "3castor" "$Date: 2006/01/26 15:36:21 $" CASTOR "Cns Library Functions"
.SH NAME
Cns_stat \- get information about a CASTOR file or directory in the name server
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "Cns_api.h"\fR
.sp
.BI "int Cns_lstat (const char *" path ,
.BI "struct Cns_filestat *" statbuf )
.sp
.BI "int Cns_stat (const char *" path ,
.BI "struct Cns_filestat *" statbuf )
.sp
.BI "int Cns_statx (const char *" path ,
.BI "struct Cns_fileid *" file_uniqueid ,
.BI "struct Cns_filestat *" statbuf )
.SH DESCRIPTION
.B Cns_stat
gets information about a CASTOR file or directory in the name server.
.LP
.B Cns_lstat
is identical to
.B Cns_stat
except for symbolic links.
.B Cns_lstat
retrieves information about the link itself, while
.B Cns_stat
gets information about the file/directory referenced by the symbolic link.
.LP
.B Cns_statx
returns the same information plus the file unique id (name server hostname
and file id).
The file can be specified by
.I path
name or by
.IR file_uniqueid .
If both are given,
.I file_uniqueid
is used.
.TP
.I path
specifies the logical pathname relative to the current CASTOR directory or
the full CASTOR pathname.
.LP
The structure pointed to by
.I statbuf
contains the following members:
.RS
u_signed64	fileid;		/* entry unique identifier */
.br
mode_t		filemode;	/* see below */
.br
int		nlink;		/* number of files in a directory */
.br
uid_t		uid;
.br
gid_t		gid;
.br
u_signed64	filesize;
.br
time_t		atime;		/* last access to file */
.br
time_t		mtime;		/* last file modification */
.br
time_t		ctime;		/* last metadata modification */
.br
short		fileclass;
.br
char		status;		/* '-' --> online, 'm' --> migrated */
.RE
.LP
filemode is constructed by OR'ing the bits defined in
.RB < sys/stat.h >
under Unix or \fB "statbits.h"\fR under Windows/NT:
.sp
.RS
.B S_IFLNK	0xA000		
symbolic link
.br
.B S_IFREG	0x8000		
regular file
.br
.B S_IFDIR	0x4000		
directory
.br
.B S_ISUID	0004000		
set user ID on execution
.br
.B S_ISGID	0002000		
set group ID on execution
.br
.B S_ISVTX	0001000		
sticky bit
.br
.B S_IRUSR	0000400		
read by owner
.br
.B S_IWUSR	0000200		
write by owner
.br
.B S_IXUSR	0000100		
execute/search by owner
.br
.B S_IRGRP	0000040		
read by group
.br
.B S_IWGRP	0000020		
write by group
.br
.B S_IXGRP	0000010		
execute/search by group
.br
.B S_IROTH	0000004		
read by others
.br
.B S_IWOTH	0000002		
write by others
.br
.B S_IXOTH	0000001		
execute/search by others
.RE
.SH RETURN VALUE
This routine returns 0 if the operation was successful or -1 if the operation
failed. In the latter case,
.B serrno
is set appropriately.
.SH ERRORS
.TP 1.3i
.B ENOENT
The named file/directory does not exist or is a null pathname.
.TP
.B EACCES
Search permission is denied on a component of the
.I path
prefix.
.TP
.B EFAULT
.I path
or
.I statbuf
is a NULL pointer.
.TP
.B ENOTDIR
A component of
.I path
prefix is not a directory.
.TP
.B ENAMETOOLONG
The length of
.I path
exceeds
.B CA_MAXPATHLEN
or the length of a
.I path
component exceeds
.BR CA_MAXNAMELEN .
.TP
.B SENOSHOST
Host unknown.
.TP
.B SENOSSERV
Service unknown.
.TP
.B SECOMERR
Communication error.
.TP
.B ENSNACT
Name server is not running or is being shutdown.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cns_chdir(3) ,
.BR Cns_chmod(3) ,
.BR Cns_chown(3) ,
.B Cns_creat(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


ns/Cns_stat.c
/*
 * Copyright (C) 1999-2005 by CERN/IT/PDP/DM
 * All rights reserved
 */

/* Cns_statx - get information about a file or a directory and
   return unique fileid */

#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include "marshall.h"
#include "Cns_api.h"
#include "Cns.h"
#include "serrno.h"

/* Cns_lstat - get information about a symbolic link */

int
Cns_lstat(const char *path, struct Cns_filestat *statbuf)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int msglen;
  char *q;
  char *rbp;
  char repbuf[57];
  char *sbp;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;
  u_signed64 zero = 0;

  strcpy (func, "Cns_lstat");
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if (! path || ! statbuf) {
    serrno = EFAULT;
    return (-1);
  }

  if (strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }

  if (Cns_selectsrvr (path, thip->server, server, &actual_path))
    return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC2);
  marshall_LONG (sbp, CNS_LSTAT);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  marshall_HYPER (sbp, zero);
  marshall_STRING (sbp, actual_path);

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, server, sendbuf, msglen, repbuf, sizeof(repbuf));
  if (c == 0) {
    rbp = repbuf;
    unmarshall_HYPER (rbp, statbuf->fileid);
    unmarshall_WORD (rbp, statbuf->filemode);
    unmarshall_LONG (rbp, statbuf->nlink);
    unmarshall_LONG (rbp, statbuf->uid);
    unmarshall_LONG (rbp, statbuf->gid);
    unmarshall_HYPER (rbp, statbuf->filesize);
    unmarshall_TIME_T (rbp, statbuf->atime);
    unmarshall_TIME_T (rbp, statbuf->mtime);
    unmarshall_TIME_T (rbp, statbuf->ctime);
    unmarshall_WORD (rbp, statbuf->fileclass);
    unmarshall_BYTE (rbp, statbuf->status);
  }
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}

int
Cns_statx(const char *path, struct Cns_fileid *file_uniqueid, struct Cns_filestat *statbuf)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int msglen;
  char *q;
  char *rbp;
  char repbuf[57];
  char *sbp;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;
  u_signed64 zero = 0;

  strcpy (func, "Cns_stat");
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if (! path || ! statbuf || ! file_uniqueid) {
    serrno = EFAULT;
    return (-1);
  }

  if (strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }

  if (file_uniqueid && *file_uniqueid->server)
    strcpy (server, file_uniqueid->server);
  else
    if (Cns_selectsrvr (path, thip->server, server, &actual_path))
      return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC);
  marshall_LONG (sbp, CNS_STAT);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  if (*file_uniqueid->server) {
    marshall_HYPER (sbp, file_uniqueid->fileid);
    marshall_STRING (sbp, "");
  } else {
    marshall_HYPER (sbp, zero);
    marshall_STRING (sbp, actual_path);
  }

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, server, sendbuf, msglen, repbuf, sizeof(repbuf));
  if (c == 0) {
    rbp = repbuf;
    unmarshall_HYPER (rbp, statbuf->fileid);
    unmarshall_WORD (rbp, statbuf->filemode);
    unmarshall_LONG (rbp, statbuf->nlink);
    unmarshall_LONG (rbp, statbuf->uid);
    unmarshall_LONG (rbp, statbuf->gid);
    unmarshall_HYPER (rbp, statbuf->filesize);
    unmarshall_TIME_T (rbp, statbuf->atime);
    unmarshall_TIME_T (rbp, statbuf->mtime);
    unmarshall_TIME_T (rbp, statbuf->ctime);
    unmarshall_WORD (rbp, statbuf->fileclass);
    unmarshall_BYTE (rbp, statbuf->status);

    strcpy (file_uniqueid->server, server);
    file_uniqueid->fileid = statbuf->fileid;
  }
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}

/* Cns_stat - get information about a file or a directory */

int
Cns_stat(const char *path, struct Cns_filestat *statbuf)
{
  struct Cns_fileid file_uniqueid;

  memset ((void *) &file_uniqueid, 0, sizeof(struct Cns_fileid));
  return (Cns_statx (path, &file_uniqueid, statbuf));
}

/* Cns_statcs - get information about a file or a directory as Cns_stat*/

int
Cns_statcs(const char *path, struct Cns_filestatcs *statbuf)
{
  struct Cns_fileid file_uniqueid;

  memset ((void *) &file_uniqueid, 0, sizeof(struct Cns_fileid));
  return (Cns_statcsx (path, &file_uniqueid, statbuf));
}

/*
 * This is the same function as Cns_statx, but has Cns_filestatcs structure as parameter.
 * Cns_filestatcs has additional fields:  csumtype, csumvalue
 */
int
Cns_statcsx(const char *path, struct Cns_fileid *file_uniqueid, struct Cns_filestatcs *statbuf)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int msglen;
  char *q;
  char *rbp;
  char repbuf[93];
  char *sbp;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;
  u_signed64 zero = 0;

  strcpy (func, "Cns_statcs");
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if (! path || ! statbuf || ! file_uniqueid) {
    serrno = EFAULT;
    return (-1);
  }

  if (strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }

  if (file_uniqueid && *file_uniqueid->server)
    strcpy (server, file_uniqueid->server);
  else
    if (Cns_selectsrvr (path, thip->server, server, &actual_path))
      return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC);
  marshall_LONG (sbp, CNS_STATCS);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  if (*file_uniqueid->server) {
    marshall_HYPER (sbp, file_uniqueid->fileid);
    marshall_STRING (sbp, "");
  } else {
    marshall_HYPER (sbp, zero);
    marshall_STRING (sbp, actual_path);
  }

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, server, sendbuf, msglen, repbuf, sizeof(repbuf));
  if (c == 0) {
    rbp = repbuf;
    unmarshall_HYPER (rbp, statbuf->fileid);
    unmarshall_WORD (rbp, statbuf->filemode);
    unmarshall_LONG (rbp, statbuf->nlink);
    unmarshall_LONG (rbp, statbuf->uid);
    unmarshall_LONG (rbp, statbuf->gid);
    unmarshall_HYPER (rbp, statbuf->filesize);
    unmarshall_TIME_T (rbp, statbuf->atime);
    unmarshall_TIME_T (rbp, statbuf->mtime);
    unmarshall_TIME_T (rbp, statbuf->ctime);
    unmarshall_WORD (rbp, statbuf->fileclass);
    unmarshall_BYTE (rbp, statbuf->status);
    unmarshall_STRING (rbp, statbuf->csumtype);
    unmarshall_STRING (rbp, statbuf->csumvalue);

    strcpy (file_uniqueid->server, server);
    file_uniqueid->fileid = statbuf->fileid;
  }
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}

int
Cns_statg(const char *path, const char *guid, struct Cns_filestatg *statbuf)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int msglen;
  char *q;
  char *rbp;
  char repbuf[130];
  char *sbp;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;

  strcpy (func, "Cns_statg");
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if ((! path && ! guid) || ! statbuf) {
    serrno = EFAULT;
    return (-1);
  }

  if (path && strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }
  if (guid && strlen (guid) > CA_MAXGUIDLEN) {
    serrno = EINVAL;
    return (-1);
  }

  if (path && Cns_selectsrvr (path, thip->server, server, &actual_path))
    return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC);
  marshall_LONG (sbp, CNS_STATG);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  if (path) {
    marshall_STRING (sbp, actual_path);
  } else {
    marshall_STRING (sbp, "");
  }
  if (guid) {
    marshall_STRING (sbp, guid);
  } else {
    marshall_STRING (sbp, "");
  }

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, path ? server : NULL, sendbuf, msglen, repbuf, sizeof(repbuf));
  if (c == 0) {
    rbp = repbuf;
    unmarshall_HYPER (rbp, statbuf->fileid);
    unmarshall_STRING (rbp, statbuf->guid);
    unmarshall_WORD (rbp, statbuf->filemode);
    unmarshall_LONG (rbp, statbuf->nlink);
    unmarshall_LONG (rbp, statbuf->uid);
    unmarshall_LONG (rbp, statbuf->gid);
    unmarshall_HYPER (rbp, statbuf->filesize);
    unmarshall_TIME_T (rbp, statbuf->atime);
    unmarshall_TIME_T (rbp, statbuf->mtime);
    unmarshall_TIME_T (rbp, statbuf->ctime);
    unmarshall_WORD (rbp, statbuf->fileclass);
    unmarshall_BYTE (rbp, statbuf->status);
    unmarshall_STRING (rbp, statbuf->csumtype);
    unmarshall_STRING (rbp, statbuf->csumvalue);
  }
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}
